[FEAT] 로그인 API

westzeroright opened the pull request:
### 개요
close #40 

### 작업사항
- 로그인 기능을 구현하였습니다.
    - 로그인을 하면 토큰에 memberId와 role 그리고 만료시간을 담아 accessToken과 refreshToken을 응답으로 반환합니다.
    - accessToken 만료시 refreshToken을 가지고 토큰을 재발급합니다.
- 인가 작업을 설정하였습니다. `JwtFilter`
- 로그인 이후 사용자의 정보를 가져올 때는 `SecurityUtils`를 사용하시면 됩니다.
  ```java
  Long userId = SecurityUtils.getCurrentMemberId();
  ```
- 로그인 기능 통합 테스트를 작성하였습니다.
- 이벤트 통합 테스트에 인증 과정을 추가하였습니다.
- 인증 예외 처리 - AuthenticationEntryPoint, 인가 예외 처리 - AccessDeniedHandler

### 변경로직
- `/v1/auth/**` url과 소속대학/학과 조회 url을 제외하고는 인증 후 인가를 통해서만 접근이 가능합니다.
- docker-compose에 redis 설정이 추가되었습니다.

### 리뷰어에게
- 현재 승인대기 과정이 없기에 권한이 guest일때 manager의 api를 사용할 수 있도록 되어 있습니다. 추후 변경
- 그리고 추후 토큰의 role과 db의 role이 다른 점을 고려해야합니다.

### 테스트 결과
- 전체 테스트 결과
   <img width="523" alt="image" src="https://github.com/user-attachments/assets/5d9f0d7c-aebb-45fc-853a-1d57be14145c" />

**추가된 테스트케이스**
- 로그인 API 컨트롤러 통합테스트
<img width="258" alt="image" src="https://github.com/user-attachments/assets/58caa3ca-d142-4e72-8de9-d5d53bde4b06" />
<img width="317" alt="image" src="https://github.com/user-attachments/assets/92da5a1f-09f9-4e5f-b349-66752c677c37" />
<img width="378" alt="image" src="https://github.com/user-attachments/assets/614d5c2e-6796-4947-9d5a-d8181786f313" />
<img width="492" alt="image" src="https://github.com/user-attachments/assets/88d4002f-0dab-4945-8d01-2c1ec188c756" />
<img width="328" alt="image" src="https://github.com/user-attachments/assets/c898011a-2b4b-458f-bb91-a2146d993657" />
<img width="320" alt="image" src="https://github.com/user-attachments/assets/ca8c1e41-bd7a-40e7-8067-c9a5259e3e8d" />

- 로그인 요청 본문 검증 테스트
<img width="360" alt="image" src="https://github.com/user-attachments/assets/07b948a6-7369-457f-8185-24336858dd07" />

- 레디스용 TokenRepository 테스트
<img width="522" alt="image" src="https://github.com/user-attachments/assets/3da2d388-dfba-453c-bf83-df7e8bfe8550" />

### 체크리스트
- [x] assignees 설정
- [x] reviewers 설정
- [x] label 설정
- [x] 브랜치 방향 확인


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **새로운 기능**
	- 사용자 인증 프로세스가 강화되어 원활한 로그인 및 토큰 재발급 기능이 추가되었습니다.
	- Redis 연동 및 JWT 기반 인증 토큰 관리 기능이 도입되어 보안성과 성능이 향상되었습니다.
	- 인증 관련 쿠키 기반 토큰 처리 및 보안 설정이 개선되어 권한 없는 접근과 인증 실패에 대한 대응이 체계화되었습니다.
	- 관리자 및 게스트 권한에 따른 API 접근 제어가 강화되었습니다.
	- 인증 관련 API 문서 및 예외 가이드가 업데이트되었습니다.
- **버그 수정**
	- 유효하지 않거나 만료된 토큰 사용 시 적절한 오류 메시지와 상태 코드가 반환되도록 개선되었습니다.
- **문서**
	- 로그인 및 토큰 재발급 관련 API 문서와 예외 처리 문서가 추가 및 보완되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

westzeroright committed:
chore: cors에 개발서버 주소 추가

westzeroright committed:
docs: 문서 제목 인증인가 -> 인가로 수정

westzeroright committed:
refactor: manager 회원가입 요청 본문 스키마 추가

westzeroright committed:
chore: redis와 jwt 의존성 추가

westzeroright committed:
feat: 로그인 API 구현

westzeroright committed:
feat: reissue 구현

westzeroright committed:
docs: 로그인 API 문서화

westzeroright committed:
test: 로그인 API 통합테스트 작성

westzeroright committed:
chroe: redis 설정 파일 작성

westzeroright committed:
feat: 필터에서 reissue 요청시 리프레시 토큰 검증

westzeroright committed:
feat: 인가작업 설정

westzeroright committed:
feat: 로그인한 유저 정보 얻기 구현

westzeroright committed:
refactor: refresh토큰 검증 위치 변경

westzeroright committed:
test: 이벤트 테스트코드 인증 통과

westzeroright committed:
test: 만료된 토큰 예외 수정

리프레시 토큰을 레포지토리에 저장했어야함

westzeroright committed:
refactor: 리프레시 토큰을 헤더에 담음

westzeroright committed:
Merge branch 'dev' into feat/#40/login

westzeroright commented on src/main/resources/application.yml:
```diff
@@ -101,6 +113,9 @@ spring:
     init:
       mode: never
 
+jwt:
+  secret: ${JWT_REFRESH_SECRET_KEY:dGVzdC1yZWZyZXNoLXNlY3JldC1rZXktdGVzdC1yZWZyZXNoLXNlY3JldC1rZXktdGVzdC1yZWZyZXNoLXNlY3JldC1rZXktdGVzdC1yZWZyZXNo}
```
그거 테스트코드용임

westzeroright replied:
아니야 너 맞말이야

Profile-exe commented:
@coderabbitai 현재 로그인 로직은 응답 본문으로 토큰을 주고있는데, Cookie를 사용해서 `accessToken`, `refreshToken`을 담아서 응답하도록 변경하려면 로그인 로직이 어떻게 바뀌어야 하지?

westzeroright committed:
refactor: 토큰을 쿠키로 보냄

westzeroright committed:
refactor: 사용하지 않는 코드 제거

westzeroright committed:
feat: 로그인 요청 본문 스키마 추가

westzeroright committed:
refacotr: CookieUtil 클래스

- sameSite 옵션도 추가하였습니다.

westzeroright committed:
refacotr: 메서드명 추출

westzeroright committed:
feat: 필터 예외 처리

westzeroright committed:
test: 액세스토큰이 null일 때

Profile-exe commented on src/main/resources/application.yml:
```diff
@@ -61,6 +61,18 @@ decorator:
+
+  data:
+    redis:
+      host: redis
+      port: 6379
```
환경변수로 관리하는게 좋아보이네요

westzeroright replied:
굿굿 중간끝나고 바로 할게요..!

westzeroright committed:
refactor: redis 설정 변수 환경변수로 관리

westzeroright committed:
feat: 인증 실패 시 예외 처리

- 스프링시큐리티에서 자동으로 이메일/비밀번호 검증
- 필터에서 발생하는 예외가 컨트롤러로 전달되지 않음
- 비즈니스익셉션으로 처리

westzeroright committed:
feat: SecurityConfig에 AuthenticationManager 설정 추가

- BCryptPasswordEncoder를 사용하고
- 구현한 CustomUserDetailService를 주입함

westzeroright committed:
test: 인증이 실패하는 경우 테스트코드 작성

- 이메일과 비밀번호 각각 잘못 입력하는 인증 실패 예외를 마주합니다.

westzeroright committed:
chore: access_token 만료시간 재설정

westzeroright committed:
style: spotless

westzeroright committed:
feat: 인가 예외 처리 핸들러 CustomAccessDeniedHandler

westzeroright committed:
refactor: contentType 상수 처리

westzeroright committed:
feat: 시큐리티 설정에 인가 핸들러 추가

westzeroright committed:
feat: 인증 api는 JwtFilter를 거치지 않도록 함

westzeroright committed:
refactor: 인가 예외 응답 커스텀

westzeroright committed:
refactor: 테스트메서드명 수정

USER_로그인시라는 문맥이 MANAGER_로그인이 존재한다고 추측할 수 있는데 그런 건 없고 두 권한 모두 하나의 엔드포인트에서 로그인을 수행함

westzeroright committed:
test: 로그인 요청 본문 검증 테스트

westzeroright committed:
docs: 로그인 API 예외 문서 작성

- 기존에 토큰 관련 예외는 로그인할 때 발생하는 것이 아니라 이후 작업과 reissue시 발생하는 거라 제거하였습니다.

westzeroright committed:
docs: reissue API 예외 문서 작성

westzeroright committed:
refacotr: 인증 실패 예외를 서비스단에서 잡지 않고 EntryPoint에서 중앙 관리하도록 함

- 추가로 리프레스토큰이 널이면 예외던지도록 함

westzeroright committed:
feat: 예외 발생시 필터체인 중단 및 비즈니스 예외를 제외한 예외를 잡음

westzeroright committed:
test: 비밀번호 검증 테스트시 여러 경우 추가

westzeroright committed:
test: 존재하지 않는 회원 로그인 응답 수정

westzeroright committed:
refactor: 응답을 JSON 형식으로 보내도록 수정

- 인증 관련 예외는 EntryPoint에서 중앙관리되고 있으나
- JwtFilter에서 reqeust.setAttribute로 설정한 예외를 EntryPoint에서 못 받을 수 있으므로
- 응답 구조 형식 코드가 필요함
- 이를 EntryPoint와 동일하게 함

westzeroright committed:
refactor: ContentType 상수 처리

westzeroright committed:
refactor: Spring의 AntPathMatcher를 사용하여 코드 간결화

westzeroright committed:
refactor: ObjectMapper를 new보다 주입받아서 사용

westzeroright committed:
feat: 도커컴포즈파일에 레디스 환경변수 추가

westzeroright committed:
refactor:  설정 파일에서 레디스 환경 변수 위치 수정

westzeroright committed:
feat: redis 설정

westzeroright committed:
test: 레디스레포지토리 crud 테스트코드

Profile-exe reviewed (approved):
수고하셨습니다 👍
